AX: figcaption should only contribute to img accname when it lacks other labelling techniques

rahimabdi opened the pull request:
#### 6bd07e47c0a5e45c5dff17579ae09909a586db74
<pre>
AX: figcaption should only contribute to img accname when it lacks other labelling techniques
<a href="https://bugs.webkit.org/show_bug.cgi?id=292490">https://bugs.webkit.org/show_bug.cgi?id=292490</a>
<a href="https://rdar.apple.com/150597445">rdar://150597445</a>

Reviewed by Tyler Wilcock.

The &lt;figure&gt; and &lt;img&gt; elements&apos; accname calculation have been updated in html-aam. Specifically, &lt;figcaption&gt; only participates in accname for
&lt;img&gt; elements (with a shared &lt;figure&gt; ancestor) that lack other labelling techniques (e.g., alt attribute, ARIA, title attribute).

This PR implements html-aam&apos;s updated &lt;figure&gt; and &lt;img&gt; calculation, and also removes the &lt;figcaption&gt; AXObjectCache::addLabelForRelation()
labelFor relation for &lt;figure&gt; because it is no longer applicable and handled elsewhere. It also updates the
LayoutTests/.../accessibility/ios-simulator/figure-element.html test to fix a missing null check which caused an invalid failure. To simplify the
check for common types of flow content for &lt;img&gt; accname computation, the static helper function isFlowContent() has been added in AccessibilityNodeObject.cpp.

* LayoutTests/imported/w3c/web-platform-tests/html-aam/figure-name-no-figcaption.tentative-expected.txt: Updated test expectations.
* LayoutTests/accessibility/mac/figure-element-expected.txt: Updated test expectations.
* LayoutTests/accessibility/ios-simulator/figure-element.html: Fixed lack of null check for titleUIElement.
* LayoutTests/accessibility/ios-simulator/figure-element.txt: Updated test expectations.
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::addLabelForRelation):
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::alternativeText const):
(WebCore::AccessibilityNodeObject::captionForFigure const):
(WebCore::isFlowContent):
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::computeIsIgnored const):

Canonical link: <a href="https://commits.webkit.org/295746@main">https://commits.webkit.org/295746@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5461d04e54d56eb6f25c6cea1b70048950db5f1e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/105995 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/25744 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/16138 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/111191 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/56591 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/108034 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/26394 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/34247 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/80515 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/108999 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/20821 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/95657 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/60837 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/20430 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/13754 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/56029 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/90200 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/13790 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/114045 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/33133 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/24446 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/89594 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/33497 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/91885 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/89277 "Found 101 new API test failures: /TestWebKit:WebKit.PageLoadDidChangeLocationWithinPage, /TestWebKit:WebKit.AboutBlankLoad, /TestWebKit:WebKit.PreventEmptyUserAgent, /TestWebKit:WebKit.PrivateBrowsingPushStateNoHistoryCallback, /WebKitGTK/TestInputMethodContext:/webkit/WebKitInputMethodContext/content-type, /WebKitGTK/TestUIClient:/webkit/WebKitWebView/pointer-lock-permission-request, /TestWebKit:WebKit.HitTestResultNodeHandle, /TestWebKit:WebKit2TextFieldBeginAndEditEditingTest.TextFieldDidBeginShouldNotBeDispatchedForAlreadyFocusedField, /TestWebKit:WebKit.Find, /WebKitGTK/TestWebProcessExtensions:/webkit/WebKitWebProcessExtension/dom-input-element-is-user-edited ... (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22759 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/34143 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/11957 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/28678 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/33058 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/38470 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/32804 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/36153 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/34402 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -1825,11 +1825,46 @@ void AccessibilityNodeObject::alternativeText(Vector<AccessibilityText>& textOrd
+                    // Iterate over the direct children of the <img>'s ancestor <figure> for any common
+                    // flow content, including non-whitespace text nodes.
+                    for (auto* node = ancestor->firstChild(); node; node = node->nextSibling()) {
+                        if (auto* element = dynamicDowncast<Element>(node); element && !element->hasTagName(figcaptionTag) && !element->hasTagName(imgTag)) {
```
I think you want HTMLElement here. Otherwise `unknownNamespace:canvas` would also end up matching, no?

rahimabdi replied:
Updated to `<HTMLElement>`.

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -1825,11 +1825,46 @@ void AccessibilityNodeObject::alternativeText(Vector<AccessibilityText>& textOrd
+        const AtomString& title = image->attributeWithoutSynchronization(titleAttr);
+
+        if (alt.isEmpty() && title.isEmpty()) {
+            for (auto* ancestor = this->node()->parentNode(); ancestor; ancestor = ancestor->parentNode()) {
```
I don't think we need to call `this->node()` again, we have a `node` local variable above that we can use.

I think we want to store ancestor in a `RefPtr` rather than a raw pointer to ensure it stays alive while we use it below. ([documentation](https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines#when-calling-a-non-trivial-member-function-on-an-object-hold-a-smart-pointer-to-the-object-on-the-stack))

rahimabdi replied:
Got it, replaced with the existing `node` instance.

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -1825,11 +1825,46 @@ void AccessibilityNodeObject::alternativeText(Vector<AccessibilityText>& textOrd
+
+        if (alt.isEmpty() && title.isEmpty()) {
+            for (auto* ancestor = this->node()->parentNode(); ancestor; ancestor = ancestor->parentNode()) {
+                if (auto* element = dynamicDowncast<Element>(ancestor); element && element->hasTagName(figureTag)) {
```
This can be a raw pointer since it's just a typecast.

But unrelatedly, I think it would be more succinct to write:

```cpp
if (auto* element = dynamicDowncast<HTMLFigureElement>(ancestor))
```

annevk replied:
I was thinking that, but `HTMLFigureElement` is not a thing.

rahimabdi replied:
I spent an embarrassing amount of time before I realized this as well 😅

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -1825,11 +1825,46 @@ void AccessibilityNodeObject::alternativeText(Vector<AccessibilityText>& textOrd
+                    bool figureHasFlowContent = false;
+                    // Iterate over the direct children of the <img>'s ancestor <figure> for any common
+                    // flow content, including non-whitespace text nodes.
+                    for (auto* node = ancestor->firstChild(); node; node = node->nextSibling()) {
```
Probably want a RefPtr here too.

rahimabdi replied:
Updated.

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -1825,11 +1825,46 @@ void AccessibilityNodeObject::alternativeText(Vector<AccessibilityText>& textOrd
+                                || element->hasTagName(pictureTag)
+                                || element->hasTagName(pTag)
+                                || element->hasTagName(spanTag)
+                                || element->hasTagName(ulTag);
```
I think it would be more efficient to retrieve the tag name once, then use that in a `switch` or a series of `if-else-if` statements. Otherwise I think we're constantly re-reading the same tag name (of `element`) over and over.

https://github.com/WebKit/WebKit/blob/0874d0175d5304debe679e48325689a39a48213d/Source/WebCore/accessibility/AccessibilityObject.cpp#L3845

This is an example, though the type of `tag` should be a `const AtomString&` rather than an `AtomString`, as the latter does an unnecessary copy.

rahimabdi replied:
Thanks, opted for `if`/`else` (don't think I need `if-else-if` since I'm evaluating a single condition).

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -1825,11 +1825,46 @@ void AccessibilityNodeObject::alternativeText(Vector<AccessibilityText>& textOrd
+                    }
+                    // If the <figcaption> is non-empty and no flow content is present in the <figure>,
+                    // the <img> derives accname from the <figcaption>.
+                    if (auto* figureObject = objectCache->getOrCreate(ancestor); figureObject && figureObject->isFigureElement()) {
```
RefPtr over auto* here too. And maybe add the `objectCache` null check, e.g.

```cpp
RefPtr figureObject = objectCache ? objectCache->getOrCreate(ancestor) : nullptr;
if (figureObject && figureObject->isFigureElement()) {
```

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -1825,11 +1825,46 @@ void AccessibilityNodeObject::alternativeText(Vector<AccessibilityText>& textOrd
+                    if (auto* figureObject = objectCache->getOrCreate(ancestor); figureObject && figureObject->isFigureElement()) {
+                        auto* figcaptionAXNodeObject = dynamicDowncast<AccessibilityNodeObject>(*figureObject)->captionForFigure();
+                        if (figcaptionAXNodeObject && !figureHasFlowContent && !figcaptionAXNodeObject->isHidden())
+                            textOrder.append(AccessibilityText(accessibleNameForNode(*figcaptionAXNodeObject->node()), AccessibilityTextSource::Alternative));
```
Couple thoughts here:

1. The only time this block does something is when `!figureHasFlowContent`. We should check that first before doing `AXObjectCache::getOrCreate`
2. figcaptionAXNodeObject is a bit of a long name. I think since it's used in such a small block, calling it just `caption` is fine.
3. In this case, because you don't check the result of the `dynamicDowncast`, you may as well use `downcast`. The difference is that the latter will cause a crash if the type isn't what is expected, whereas `dynamicDowncast` just returns `nullptr`. But you're not null-checking the result anyways, which is probably fine (we know this object is associated with a figure element, which must be an `AccessibilityNodeObject`).

rahimabdi replied:
> Couple thoughts here:
> 
> 1. The only time this block does something is when `!figureHasFlowContent`. We should check that first before doing `AXObjectCache::getOrCreate`

Agreed, updated.

> 2. figcaptionAXNodeObject is a bit of a long name. I think since it's used in such a small block, calling it just `caption` is fine.

Changed.

> 3. In this case, because you don't check the result of the `dynamicDowncast`, you may as well use `downcast`. The difference is that the latter will cause a crash if the type isn't what is expected, whereas `dynamicDowncast` just returns `nullptr`. But you're not null-checking the result anyways, which is probably fine (we know this object is associated with a figure element, which must be an `AccessibilityNodeObject`).

Updated, good to know! Thanks.

twilco commented on Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
```diff
@@ -1218,6 +1218,13 @@ bool AccessibilityRenderObject::computeIsIgnored() const
+        for (auto* ancestor = renderer(); ancestor; ancestor = ancestor->parent()) {
+            if (auto* element = dynamicDowncast<HTMLElement>(ancestor->node()); element && element->hasTagName(figureTag))
+                return false;
+        }
```
`::computeIsIgnored` is one of our hottest codepaths, and this adds an O(n) ancestry traversal. I think it's OK because this is limited to things that are `isImage`. I think we could make it a tiny bit cheaper by directly traversing the DOM ancestry via `parentNode` rather than traversing the render tree ancestry, then getting the DOM node at every step. Also, this check would probably be more efficient (and more succinct):

```cpp
if (auto* element = dynamicDowncast<HTMLElement>(ancestor->node()); element && element->hasTagName(figureTag))
```

rahimabdi replied:
Is there a proposed change here? (Apologies, it looks the same).

rahimabdi replied:
This ended up being:

`if (auto* element = dynamicDowncast<HTMLElement>(ancestor.get()); element && element->hasTagName(figureTag))`

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -1825,11 +1825,46 @@ void AccessibilityNodeObject::alternativeText(Vector<AccessibilityText>& textOrd
-        AccessibilityObject* captionForFigure = this->captionForFigure();
-        if (captionForFigure && !captionForFigure->isHidden())
-            textOrder.append(AccessibilityText(accessibleNameForNode(*captionForFigure->node()), AccessibilityTextSource::Alternative));
+    if (auto* image = dynamicDowncast<HTMLImageElement>(*node); image && objectCache) {
```
I would remove this null check of `objectCache` since the check is so far removed from where we actually dereference it. It would be better to null check it closer to the first dereference.

twilco commented on LayoutTests/accessibility/ios-simulator/figure-element.html:
```diff
@@ -65,9 +65,14 @@
+                    shouldBeTrue("titleUIElement.isEqual(figureCaption)");
+                } else {
+                    shouldBeTrue("isIgnored(titleUIElement) && isIgnored(figureCaption)");
+                }
```
I'd probably write it this way instead (I don't think we need to null-check `figureCaption` but maybe I'm wrong):

```js
var titleEqualsCaption = titleUIElement && titleUIElement.isEqual(figureCaption);
shouldBeTrue("titleEqualsCaption || (isIgnored(titleUIElement) && isIgnored(figureCaption))");
```

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -1818,11 +1818,44 @@ void AccessibilityNodeObject::alternativeText(Vector<AccessibilityText>& textOrd
-        AccessibilityObject* captionForFigure = this->captionForFigure();
-        if (captionForFigure && !captionForFigure->isHidden())
-            textOrder.append(AccessibilityText(accessibleNameForNode(*captionForFigure->node()), AccessibilityTextSource::Alternative));
+    if (auto* image = dynamicDowncast<HTMLImageElement>(*node); image) {
```
The `; image` is not necessary — `if (auto* foo = ...)` inherently only executes if `foo` is non-null.

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -1818,11 +1818,44 @@ void AccessibilityNodeObject::alternativeText(Vector<AccessibilityText>& textOrd
+        const AtomString& alt = image->attributeWithoutSynchronization(altAttr);
+        const AtomString& title = image->attributeWithoutSynchronization(titleAttr);
+
+        if (alt.isEmpty() && title.isEmpty()) {
```
It would be slightly more efficient to write this, ensuring we only retrieve title if `alt` is empty:

```cpp
const AtomString& alt = image->attributeWithoutSynchronization(altAttr);
if (alt.isEmpty() && image->attributeWithoutSynchronization(titleAttr).isEmpty()) {
    ...
}
```

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -1818,11 +1818,44 @@ void AccessibilityNodeObject::alternativeText(Vector<AccessibilityText>& textOrd
+                            String captionAccname = accessibleNameForNode(*caption->node());
+                            if (caption && !caption->isHidden() && !captionAccname.isEmpty())
+                                textOrder.append(AccessibilityText(captionAccname, AccessibilityTextSource::Alternative));
+                        }
```
A bit more terse way to write this block would be:

```cpp
RefPtr figureObject = objectCache ? objectCache->getOrCreate(*ancestor) : nullptr;
RefPtr caption = figureObject && figureObject->isFigureElement() ? dynamicDowncast<AccessibilityNodeObject>(figureObject)->captionForFigure() : nullptr;
if (caption && !caption->isHidden()) {
    RefPtr captionNode = caption->node();
    if (String captionAccname = captionNode ? accessibleNameForNode(*captionNode) : emptyString(); !captionAccname.isEmpty())
        textOrder.append(AccessibilityText(WTFMove(captionAccname), AccessibilityTextSource::Alternative));
}
```

This also fixes several issues:
* `auto* caption` needs to be a RefPtr since it is [used non-trivially](https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines#when-calling-a-non-trivial-member-function-on-an-object-hold-a-smart-pointer-to-the-object-on-the-stack)
* `caption->node()` was missing a null-check
* Generally better to use `dynamicDowncast` rather than `downcast`, even though in this it always _should_ be an `AccessibilityNodeObject`, better to fail gracefully here rather than crashing as `downcast` does.
* We can move the `captionAccname` into the `AccessibilityText` to avoid a string copy

rahimabdi replied:
Awesome, thanks!

twilco commented on Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
```diff
@@ -1218,6 +1218,14 @@ bool AccessibilityRenderObject::computeIsIgnored() const
+        // https://github.com/w3c/aria/pull/2224
+        // If the image is a descendant of a <figure>, don't ignore it due to <figure> potentially participating in its accname.
+        Node* node = m_renderer->node();
+        for (RefPtr ancestor = node->parentNode(); ancestor; ancestor = ancestor->parentNode()) {
```
Missing null-check on this node. This would work:

```cpp
auto* node = m_renderer->node();
for (RefPtr ancestor = node ? node->parentNode() : nullptr; ancestor; ancestor = ancestor->parentNode()) {
```

Raw pointer is OK for `node` since we only use it to call `parentNode`, which is a "trivial" function (inlined read of a member variable).

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -1818,11 +1818,44 @@ void AccessibilityNodeObject::alternativeText(Vector<AccessibilityText>& textOrd
+                    // flow content, including non-whitespace text nodes.
+                    for (RefPtr node = ancestor->firstChild(); node; node = node->nextSibling()) {
+                        if (auto* element = dynamicDowncast<HTMLElement>(node.get()); element && !element->hasTagName(figcaptionTag) && !element->hasTagName(imgTag)) {
+                            const AtomString& tag = element ? element->localName() : nullAtom();
```
No need to null-check `element` here. We wouldn't enter this block if element was nullptr.

rahimabdi replied:
I've removed this*** `element` null check:

```
     const AtomString& tag = element->localName();
```

twilco commented on LayoutTests/accessibility/ios-simulator/figure-element.html:
```diff
@@ -65,10 +65,9 @@
                 var figureCaption = accessibilityController.accessibleElementById("figCaption" + k);
-                shouldBeTrue("titleUIElement.isEqual(figureCaption) || (isIgnored(titleUIElement) && isIgnored(figureCaption))");
-            } else
-                shouldBeTrue("!titleUIElement");
```
This removes test coverage, doesn't it? Why do we no longer verify some state for `k < 5`?

rahimabdi replied:
This looks like a copy/paste error, fixed.

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -1817,11 +1817,43 @@ void AccessibilityNodeObject::alternativeText(Vector<AccessibilityText>& textOrd
+
+                        if (auto* text = dynamicDowncast<Text>(node.get()); text && !text->data().containsOnly<isASCIIWhitespace>())
+                            figureHasFlowContent = true;
+                    }
```
As an optimization, couldn't we `break` out of the loop when we set `figureHasFlowContent = true`?

rahimabdi replied:
Yes! Looks like I have a `break` for the outer loop, but not the inner one.

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -1817,11 +1817,43 @@ void AccessibilityNodeObject::alternativeText(Vector<AccessibilityText>& textOrd
+                    }
+                    // If no flow content is present in the <figure>, the <img> derives accname from its <figcaption>.
+                    if (!figureHasFlowContent) {
+                        RefPtr figureObject = objectCache ? objectCache->getOrCreate(*ancestor) : nullptr;
```
Is calling this local variable `figureObject` accurate? `ancestor` seems to be some arbitrary ancestor node, not necessarily a figure element one, right?

rahimabdi replied:
> Is calling this local variable figureObject accurate?

I believe so due to `ancestor` being scoped to the outer loop that searches for the image's `<figure>` ancestor (although not referenced until the `!figureHasFlowContent` check):

```
for (RefPtr ancestor = node->parentNode(); ancestor; ancestor = ancestor->parentNode()) {
...
}
```

Any ideas on improving/clarifying this?

rahimabdi replied:
Actually, here is where we get an element reference for the actual `<figure>`:

`if (auto* element = dynamicDowncast<HTMLElement>(ancestor.get()); element && element->hasTagName(figureTag))`

rahimabdi replied:
Updated this variable to `figure` (from `element`) if that's clearer?

`if (auto* figure = dynamicDowncast<HTMLElement>(ancestor.get()); figure && figure->hasTagName(figureTag))`

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -1817,11 +1817,43 @@ void AccessibilityNodeObject::alternativeText(Vector<AccessibilityText>& textOrd
+                    // If no flow content is present in the <figure>, the <img> derives accname from its <figcaption>.
+                    if (!figureHasFlowContent) {
+                        RefPtr figureObject = objectCache ? objectCache->getOrCreate(*ancestor) : nullptr;
+                        RefPtr caption = figureObject && figureObject->isFigureElement() ? dynamicDowncast<AccessibilityNodeObject>(figureObject)->captionForFigure() : nullptr;
```
My mistake as I realize this is exactly what I suggested. But if we aren't going to null-check the result of `dynamicDowncast` before dereferencing it, there's no point in using `dynamicDowncast` over `downcast`. Let's use `downcast` here.

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -1817,11 +1817,48 @@ void AccessibilityNodeObject::alternativeText(Vector<AccessibilityText>& textOrd
+                            figureHasFlowContent = true;
+                            break;
+                        }
+
```
Extra newline here.

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -1817,11 +1817,48 @@ void AccessibilityNodeObject::alternativeText(Vector<AccessibilityText>& textOrd
+                    bool figureHasFlowContent = false;
+                    // Iterate over the direct children of the <img>'s ancestor <figure> for any common
+                    // flow content, including non-whitespace text nodes.
+                    for (RefPtr node = ancestor->firstChild(); node; node = node->nextSibling()) {
```
So sorry to be annoying about naming 😅 but rather than `ancestor->firstChild()`, maybe we can use `figure->firstChild()` to be clear about what children we're iterating over (the children of the figure)?

And rather than `node`, maybe `figureChildNode`, and rather than `element`, `figureChildElement`? There are so many layers of if's and for's here that I feel it's easy to lose track of what these nodes and elements are without these more explicit names.

rahimabdi replied:
@twilco Not at all, agreed and thank you very much for the detailed feedback.

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -1817,11 +1817,48 @@ void AccessibilityNodeObject::alternativeText(Vector<AccessibilityText>& textOrd
+                        if (auto* text = dynamicDowncast<Text>(node.get()); text && !text->data().containsOnly<isASCIIWhitespace>()) {
+                            figureHasFlowContent = true;
+                            break;
+                        }
```
One thing that would make all of this more readable — what if we created a helper function? One idea:

```cpp
// This function purposely excludes <img> and <figcaption>, as it's only current purpose is to assist
// in an accname calculation.
static bool isFlowContent(Node& node)
{
    if (auto* element = dynamicDowncast<HTMLElement>(node)) {
        const AtomString& tag = element->localName();
        if (tag == blockquoteTag || tag == canvasTag || tag == codeTag || tag == divTag || tag == olTag || tag == pictureTag || tag == preTag || tag == pTag || tag == spanTag || tag == ulTag)
            return true;
    }

    auto* text = dynamicDowncast<Text>(node);
    return text && !text->data().containsOnly<isASCIIWhitespace>();
}
```

Then your calling code can be a lot simpler.

```cpp
bool figureHasFlowContent = false;
for (RefPtr figureChild = figure->firstChild(); figureChild; figureChild = figureChild->nextSibling()) {
    if (isFlowContent(*figureChild) {
        figureHasFlowContent = true;
        break;
   }
}
```

Wrote all of that in the GitHub comment pane, so there may be some inaccuracies, but hopefully gets the idea across.

rahimabdi replied:
Agreed! Thank you.

twilco reviewed (commented):
Almost there, thank you for sticking with this one, it will be a nice improvement 🙏

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -735,6 +735,27 @@ bool AccessibilityNodeObject::hasElementDescendant() const
+
+    auto* text = dynamicDowncast<Text>(node);
+    return text && !text->data().containsOnly<isASCIIWhitespace>();
+}
```
The only reason for this to be a member function (defined in the header under the `AccessibilityObject` class) is if it uses member variables or member functions itself. For example if rather than passing in a `Node&`, we used the `node()` method associated with `AccessibilityObject`, then it would make sense.

I think a better fit here would be a static function, i.e.:

```cpp
static bool isFlowContent(Node& node)
{
    ...
}
```

And then we can remove the definition in the header too.

twilco reviewed (approved):
r+ with last static function change

rahimabdi committed:
AX: figcaption should only contribute to img accname when it lacks other labelling techniques
https://bugs.webkit.org/show_bug.cgi?id=292490
rdar://150597445

Reviewed by Tyler Wilcock.

The <figure> and <img> elements' accname calculation have been updated in html-aam. Specifically, <figcaption> only participates in accname for
<img> elements (with a shared <figure> ancestor) that lack other labelling techniques (e.g., alt attribute, ARIA, title attribute).

This PR implements html-aam's updated <figure> and <img> calculation, and also removes the <figcaption> AXObjectCache::addLabelForRelation()
labelFor relation for <figure> because it is no longer applicable and handled elsewhere. It also updates the
LayoutTests/.../accessibility/ios-simulator/figure-element.html test to fix a missing null check which caused an invalid failure. To simplify the
check for common types of flow content for <img> accname computation, the static helper function isFlowContent() has been added in AccessibilityNodeObject.cpp.

* LayoutTests/imported/w3c/web-platform-tests/html-aam/figure-name-no-figcaption.tentative-expected.txt: Updated test expectations.
* LayoutTests/accessibility/mac/figure-element-expected.txt: Updated test expectations.
* LayoutTests/accessibility/ios-simulator/figure-element.html: Fixed lack of null check for titleUIElement.
* LayoutTests/accessibility/ios-simulator/figure-element.txt: Updated test expectations.
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::addLabelForRelation):
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::alternativeText const):
(WebCore::AccessibilityNodeObject::captionForFigure const):
(WebCore::isFlowContent):
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::computeIsIgnored const):

Canonical link: https://commits.webkit.org/295746@main
